oasislmf/pytools/gul/io.py

```diff
@@ -2,283 +2,351 @@
+            compute_i, compute, items_data_i, items_data, seeds, rng_index, group_id_rng_index,
+            damagecdf_i, rec_idx_ptr
+        )
 
-        if yield_event:
-            # event is fully read, append the last chunk of data to the list of this event
-            damagecdfs.append(damagecdf[:i])
-            Nbinss.append(Nbins[:i])
-            recs.append(rec[:i])
+        if cursor == 0 and len_read == 0:
+            # here valid buff > 0, but not enough data to have a full item => the stream stopped prematurely
+            raise Exception(
+                f"cdf input stream ended prematurely, event_id: {event_id}, valid_buf: {valid_buf}, buff_size: {buff_size}"
+            )
 
-            yield last_event_id, np.concatenate(damagecdfs), np.concatenate(Nbinss), np.concatenate(recs)
+        # persist the cdf records read from the stream
+        recs.append(rec[:rec_idx_ptr[-1]])
 
-            # start a new list for the new event, storing the first element
-            damagecdfs, Nbinss, recs = [damagecdf[i:i + 1]], [Nbins[i:i + 1]], [rec[i:i + 1]]
+        if yield_event:
+            # event is fully read
+            yield last_event_id, compute_i, items_data, np.concatenate(recs), rec_idx_ptr, rng_index
+
+            # init the data structures for the next event
             last_event_id = event_id
 
-        else:
-            # the current event is not finished, keep appending data about this event
-            damagecdfs.append(damagecdf[:i + 1])
-            Nbinss.append(Nbins[:i + 1])
-            recs.append(rec[:i + 1])
+            # clear data structures
+            group_id_rng_index, rec_idx_ptr = gen_structs()
+            rng_index = 0
+            damagecdf_i = 0
+            compute_i = 0
+            items_data_i = 0
+            coverages['cur_items'].fill(0)
+            recs = []
 
         # convert cursor to bytes
         cursor_buf = cursor * int32_mv.itemsize
 
-        if valid_buf == cursor_buf:
-            # this is the last cycle, all data has been read, append the last chunk of data
-            damagecdfs.append(damagecdf[:i])
-            Nbinss.append(Nbins[:i])
-            recs.append(rec[:i])
+        # move the un-read data to the beginning of the memoryview
+        mv[:valid_buf - cursor_buf] = mv[cursor_buf:valid_buf]
 
-            yield last_event_id, np.concatenate(damagecdfs), np.concatenate(Nbinss), np.concatenate(recs)
-            break
+        # update the length of the valid data
+        valid_buf -= cursor_buf
 
-        else:
-            # this is not the last cycle
-            # move the un-read data to the beginning of the memoryview
-            mv[:valid_buf - cursor_buf] = mv[cursor_buf:valid_buf]
 
-            # update the length of the valid data
-            valid_buf -= cursor_buf
+@njit(cache=True, fastmath=True)
+def insert_item_data_val(items_data, item_id, damagecdf_i, rng_index):
```

Comment:
The ordering here is not useful, we can insert directly in the cur_item position then increment cur_item

Reply:
Ok, I've removed the insert-sort.